### ISSUE TITLE
Create a Synology package

### ISSUE THREAD

**Issue by kenkendk** [duplicati/duplicati#1435]
The package format is described here:
http://forum.synology.com/wiki/index.php/Synology_package_files

**Comment by Lazarix** [comment]
I'd also like this

**Comment by paulinhoocx** [comment]
It would be very very good!

**Comment by hcjehg** [comment]
Here is a (very quick) start (duplicati2)

It is really a tgz file. Inside the tgz the files from duplicati has just been packages into package.tgz (2.0.0.93). To replace with a newer version replace the contents in the inner package.tgz

https://mail.hobsoft.net/owncloud/index.php/s/dsbE6Kd2obxzYlf

This file is for manual install

Now... It almost works. All seems to work (no log) except the Running status in package manager. I have tried nearly everything, including making fake return codes but package manager ALWAYS shows Stopped, despite Duplicati running.... Also no concept for log's yet.

I just installed the Beta Mono from Synology. There is a file missing in libs that needs to be copied from the Mono package to make it work, otherwise it works nicely.

Other small problem is that in order to make Duplicati listen on the external interface (and not only port 8200) you need to log in and change the setting (Duplicati - Options - Allow remote access)

I ssh-tunneled in to get this done, but there are perhaps a command-line option ?

Anyway, perhaps someone has a good idea. Me I am out of them for the moment....

HC - Who is looking forward to the Amazon Cloud Drive option :-D

**Comment by kenkendk** [comment]
Very nice!

There are a few options to control the behaviour of the webserver:
```
--webservice-port=8200
--webservice-interface=any
--webservice-password=secret_password
```

Beware that if you enable external access to the webservice, remote user can read at least your filenames. What is needed is a http-relay service (I think Synology has one?), such that all files under "webroot" are exposed through the normal webserver, and everything with "/control.cgi" and "/api/*" are sent to the Duplicati webservice, if you are logged in to the Synology WebUI.

**Comment by kenkendk** [comment]
According to the documentation, the status is reported by the start-stop-status script:
https://global.download.synology.com/download/Document/DeveloperGuide/DSM_Developer_Guide.pdf

It also states that you need to return 0 if the program is running, and right now you appear to return 1 when it is running.

Also, can I add the your script files to the repo, so it can be part of the auto-build process ?

**Comment by hcjehg** [comment]
for the status command exit $(( ! $? ))  this inverts the outcome of DaemonStatus(), anyway it doesn't seem to matter what return code is (at least 0 or 1) Package manager always shows Stopped. I must have done something wrong or missed a detail :-)

**Comment by hcjehg** [comment]
Add to repo, absolutely, yes please :-)

**Comment by hcjehg** [comment]
I should mention that the file is created by using content/and inspiration from other files mainly 
https://amigotechnotes.wordpress.com/2014/05/17/how-to-create-a-spk-for-synology-dsm-to-distribute-your-lamp/
and

The Mono file to be copied and or a conf file corrected is libMonoPosixHelper.so I copied it from /volume1/@appstore/Mono/usr/lib/libMonoPosixHelper.so to /lib/libMonoPosixHelper.so

That's probably not the correct solution, but it works...



Here are the commands used for creating the package - for inspiration :-)



cd "to the content you make the release zip file from"
rm package.tar ; tar -cvf package.tar *
mv package.tar "to the duplicati2 synology package directory"
cd "to the duplicati2 synology package directory" 
rm duplicati2.spk ; tar -cvf duplicati2.spk *

HC

**Comment by hcjehg** [comment]
Using this command (as root) one can convince one selve that the return code is good:

**Comment by hcjehg** [comment]
Arrgh Ctrl-Enter submits....

SYNOPKG_PKGNAME="Duplicati2" ; export SYNOPKG_PKGNAME ; /var/packages/Duplicati2/scripts/start-stop-status status ; echo "$?"

HC

**Comment by hcjehg** [comment]
And for now, the access to the interface is through TCP port 8200, not - as it would otherwise be very elegant - through the interface of the Synology...

Someone might have a good idea to integrate with WebStation ?

HC

**Comment by hcjehg** [comment]
I forgot that in order to make duplicati work with for instance Hubic you need to run the 

mono /pathtoinstalldir/mozroots.exe --import --sync

command.

This should be run during install and perhaps every now and then to make sure certificates are up to date...

HC

**Comment by hcjehg** [comment]
Bon, this morning I created a package with 2.0.0.94 and it seems to work with the status :-O

**Comment by kenkendk** [comment]
Great!

I will see if I can make a simple build process for this.

I have applied for access to the Synology store, so Duplicati can become an easy-install package at some point.

**Comment by hcjehg** [comment]
Now running 2.0.0.95 and 64bit java (installed via Java Manager) on Synology with HubiC backend.

64 bit becomes interesting if you have a large backup set (~>2.7TB) because it is not possible to assign more that 2G heap in the 32 bit version. 64bit can have a larger heap, currently running 8G.

HC

https://mail.hobsoft.net/owncloud/index.php/s/jPhXpjVhU0idsaU

**Comment by hcjehg** [comment]
Now running 2.0.0.95 on Synology DS1815+ with HubiC backend.

HC

Here is a synology package for those adventurous souls that wants to test.

https://mail.hobsoft.net/owncloud/index.php/s/jPhXpjVhU0idsaU

**Comment by kenkendk** [comment]
I assume you mean 64bit Mono?

**Comment by hcjehg** [comment]
Hello Kenneth

No I was writing this at a time where I should have had more coffe :-) I am running Duplicati, but also CrashPlan - which I hope to replace with Duplicati2. Crashplan is very memory hungry so the 32bit java installed by the Java Manager is quickly running out of space (max 2GB heap). I simply wrote that in the wrong project :-) Tomorrow morning I will make the coffee double strength :-D

By the way:

Is there any advantage in running Duplicati with 64bit Mono ? (Does any one know of a Synology package. The official pack (in Beta) is 32 bit - perhaps that will change with DSM6.0)? 

And is Duplicati2 compiled as 32 or 64 bit? (Does it even make sense?)

HC

**Comment by kenkendk** [comment]
Duplicati is architecture agnostic, so it can run on whatever the Mono/.Net runtime has.

The primary benefit from 64-bit is larger memory allocations, but I don't think you run into that with Duplicati anymore.

**Comment by hcjehg** [comment]
Speaking of memory. Does Duplicati scale well in terms of storage backed up. If one was to back up 50TB with Duplicati, how would that, in your opinion/best guess, perform with Duplicati?

HC

**Comment by kenkendk** [comment]
It should scale, but performance drops as the database grows large.
You can use a larger blocksize to reduce this problem.
For something like 50TB you could use a 10MB blocksize and 500MB volumes.
But if you have many files, you might see some slowdown from that too.

**Comment by hcjehg** [comment]
With this configuration, how much ram would you guestimate Duplicati2 would need to do a complete run?

**Comment by kenkendk** [comment]
The memory consumption should be the same. There is no extra memory usage with larger backup sizes (a lot of the complexity in the code is there only to prevent loading everything into memory).

The only thing that grows is the database. A rough estimate is that you need to store around 100 bytes pr. block (64 bytes of hash-text and some ID's). The number of blocks is roughly total-size / blocksize.

**Comment by kenkendk** [comment]
I had a look at the synology package. It seems fairly simple to get it running within the DSM webserver.
You need to set the `webui` path in the `info` file to `webroot`, and then add a `config` file inside the webroot folder. That should take care of serving all the static content.

Getting the REST api working is more challenging. I think the easiest approach would be to create a single cgi executable that forwards requests to the internal Duplicati webserver. All calls from the UI should then be "tunneled" such that the real url (e.g. /api/v1/backup) is attached as a header. This would give a single entry-point to the webserver and allow the DSM to protect against outside access.

**Comment by hcjehg** [comment]
Running with 500MB volume 10MB Block on a 16GB Ram Synology ...

System.OutOfMemoryException: Out of memory
at <0x00000> <unknown method>
at System.IO.MemoryStream.set_Capacity (Int32 value) [0x00000] in <filename unknown>:0 
at System.IO.MemoryStream.Expand (Int32 newSize) [0x00000] in <filename unknown>:0 
at System.IO.MemoryStream.Write (System.Byte[] buffer, Int32 offset, Int32 count) [0x00000] in <filename unknown>:0 
at Mono.Security.Protocol.Tls.TlsStream.Write (System.Byte[] buffer, Int32 offset, Int32 count) [0x00000] in <filename unknown>:0 
at Mono.Security.Protocol.Tls.TlsStream.Write (Byte value) [0x00000] in <filename unknown>:0 
at (wrapper remoting-invoke-with-check) Mono.Security.Protocol.Tls.TlsStream:Write (byte)
at Mono.Security.Protocol.Tls.RecordProtocol.EncodeRecord (ContentType contentType, System.Byte[] recordData, Int32 offset, Int32 count) [0x00000] in <filename unknown>:0 
at Mono.Security.Protocol.Tls.SslStreamBase.InternalBeginWrite (Mono.Security.Protocol.Tls.InternalAsyncResult asyncResult) [0x00000] in <filename unknown>:0

**Comment by hcjehg** [comment]
Synology package 2.0.0.98 available at 
https://mail.hobsoft.net/owncloud/index.php/s/OnYHoo41TlpFOnZ

**Comment by robsonsobral** [comment]
Hi!

Have you considered, guys, to submit this package to [Synocomunity](https://synocommunity.com/)?

Duplicati was considered [once](https://github.com/SynoCommunity/spksrc/issues/1927).

**Comment by mmleidl** [comment]
Hi! 
Me too, I'm trying to get duplicati on a synology NAS. Any news about a Synology Package comeing up?

**Comment by kenkendk** [comment]
@robsonsobral  I think SynoCommunity does not support the ARM based machines (which is what I have).

@mmleidl: I have not worked on it, but I am working on installers and I hope that will lead to me making the cgi proxy I mentioned. After that is done, it is really easy to make the package based on the work from @hcjehg .

**Comment by robsonsobral** [comment]
@kenkendk , you can check the architetures per model on [SynoCommunity wiki](https://github.com/SynoCommunity/spksrc/wiki/Architecture-per-Synology-model)

I have several packages from them running on a DS414. Isn't it an ARM?

**Comment by kenkendk** [comment]
@robsonsobral I misread the front page https://github.com/SynoCommunity/spksrc:
> ... Non-x86 architectures are not supported.

But that applies only to the build tools, not necessarily the packages.

**Comment by kenkendk** [comment]
I have written most of a proxy now, such that the lighttpd in Synology can serve the html files and proxy requests from authenticated users to the Duplicati.Server process. This should allow nice integration with the Duplicati UI showing up as a window in DSM.

**Comment by stuartisaac** [comment]
@hcjehg would you mind posting another package? I tried some of the links, but they appear to be dead.

**Comment by kenkendk** [comment]
The Synology packages are now part of the regular builds, and you can get the .spk file here:
https://github.com/duplicati/duplicati/releases

It has some minor issues, which are recorded here:
https://github.com/duplicati/duplicati/issues/1766

Most annoying is that the initial start of Duplicati fails, so you need to close the window and try again.
This is related to the xsrf token. Also, the requests are quite slow. On my DS211 a request takes around 16 seconds to complete, so be patient when clicking around.

The reason for the delay is the proxy logic. I could not find a reliable way to re-configure the nginx server used by DSM, so instead I use the CGI interface to pipe requests through to Duplicati. It would be faster if someone would rewrite the CGIProxyHandler in Perl I think, or figure out how to insert custom nginx configurations like DownloadStation and others do.

The upside (and reason I use the DSM webserver) is that no additional ports are required. Also, the requests are verified to make sure only a DSM administrator can query Duplicati.

**Comment by robsonsobral** [comment]
Cool, @kenkendk ! Any plans of summit it to Synocommunity as an alpha/beta/test package?

**Comment by kenkendk** [comment]
@robsonsobral That sounds like a great idea. Do you have links or experience with the proces?

**Comment by robsonsobral** [comment]
I don't have any experience on that. I'm sorry.

This is what I found:

- there's a [wiki with useful info](https://github.com/SynoCommunity/spksrc/wiki/Developers-HOW-TO), like the [ports used by Synology](https://github.com/SynoCommunity/spksrc/wiki/Synology-Used-Ports) and the [ports used by other packages](https://github.com/SynoCommunity/spksrc/wiki/SynoCommunity-Used-Ports);
- a [package request](https://github.com/SynoCommunity/spksrc/blob/master/CONTRIBUTING.md#package-requests) can be a good way of start interaction;
- every package has a directory on ` /spk` of the repository [spksrc](https://github.com/SynoCommunity/spksrc) and you need to [submit a pull request to it](https://github.com/SynoCommunity/spksrc/blob/master/CONTRIBUTING.md#pull-requests);
- there's [a example](https://github.com/SynoCommunity/spksrc/wiki/spksrc-example) on docs;

I hope they're useful.

**Comment by kenkendk** [comment]
Ok, I added an issue, lets see what happens: https://github.com/SynoCommunity/spksrc/issues/2342

**Comment by stuartisaac** [comment]
I've found with the package that drop-down boxes don't present me any options. For example, if I want to look at live log data, I go to the "Select a log level and see messages as they happen:" drop-down, but Disabled is the only option that I get. I'm using the 2.0.1.17 package. Any idea what that's about?

**Comment by stuartisaac** [comment]
I seem to have  found the issue, and I'll post the resolution here in case it helps someone else out. In addition to the issue that I reported, I was getting errors like so in the log:
```
Request for http://my_ip_address:5001/api/v1/remoteoperation/test gave error
System.Net.WebException: The remote server returned an error: (401) Unauthorized.
  at Duplicati.Library.Utility.AsyncHttpRequest+AsyncWrapper.GetResponseOrStream () <0x40b228f0 + 0x000fb> in <filename unknown>:0 
  at Duplicati.Library.Utility.AsyncHttpRequest.GetResponse () <0x40b71d00 + 0x00157> in <filename unknown>:0 
  at Duplicati.Library.JSONWebHelper.GetResponse (Duplicati.Library.Utility.AsyncHttpRequest req, System.Object requestdata) <0x40c08710 + 0x003ef> in <filename unknown>:0
```

I started poking around in the Synology security-related settings, ended up in Control Panel > Network > DSM Settings, and I unchecked "Enable HTTP/2". After saving that change, logging out, and logging in again, these errors in Duplicati went away.

**Comment by stuartisaac** [comment]
I'll add, though, that I have two remaining issues:

1. Connection tests to Amazon Cloud Drive always fail. Things seem to work despite the connection test failures, but...?
2. I can't seem to export the backup configuration successfully to file. I just get the page that says:
```
Export backup configuration
If the backup file was not downloaded automatically,
right click and choose "Save as ..."
```

I'm using Safari on OS X, and I right-click on the link and choose Download Linked File As, but the Safari download manager just gives me an error like `The requested URL was not found on this server`. For now, I've exported the command line options and can re-create things given that, but it would be nice to have the file. (I suppose I can reconstruct a file given a different one that I have stored).
